### DISCUSSION TITLE
Allow entity services to set entity target state

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#1059] [Entity Models]
## What this proposal aims to improve
- Allow us to send the target state back to other home automation systems, for example Google Home, to avoid flapping in their UIs
- Improve on how state changes are attributed to service calls; today any state change within 5 seconds of a service call is attribute to the service call, this can cause very misleading logbook entries where an automation which turns on a light may appear to sometimes turn off a light and the like

## Proposal
- Each base component, e.g. `light`, `switch`, will be able to to:
  - generate target states based on service call data
  - support matching state update with target state
- Store target state as a difference to the current state, e.g:
  - Turning on a switch which is off will generate target state `{state: on }`
  - Changing brightness of a light which is already on will generate a target state `{attributes: {brightness: 123}}`
- Target state will have an expiry time
- Target state will hold a reference to service call context
- Target state do not need to fire events because only the initiator of a service call is interested in the target state
  - Return target state and it's expiry time from service call, regardless of if service call is called blocking or not
  - Caller is responsible for reverting to the actual state if there's no state update within the expiry time (update UI, update voice assistant, etc.)

## Unclear
- Do we need to keep track of multiple pending target states for the same entity, e.g.:
  - service call 1: turn on `light.test` -> `{state: on}`
  - service call 2: set brightness to 123 `{attributes: {brightness: 123}}`
- Do we need to consider the case when there is not a 1:1 mapping between state update and service call:
  - A single service call may result in multiple state updates (`zha` is known to do this)
  - Multiple service calls may result in a single state update (this is likely to be the case for any polled integration)
- Can any state update invalidate a target state, for example if a light's target state is `{state: off }`, does a state update where brightness is changed invalidate / cancel the target state.

**Comment by MartinHjelmare** [comment]
Returning the target state from the service call sounds to conflict semantically with service response data.

**Reply by emontnemery** [reply]
Yes, the service helpers would need to be updated to return an object which holds both target state(s) and the service response data.

**Reply by MartinHjelmare** [reply]
Eg an entity service returns a dict with the entity_id as keys. Putting something else in the same dict seems confusing when using the dict.

**Reply by emontnemery** [reply]
Yes, sure, if entity target state is returned from the service call we should not mix it with the service response dictionary.

**Comment by evelant** [comment]
I think this could also be useful for improving reliablity of service calls. AFAICT right now they're "fire and forget" unless the user implements their own check/retry logic. Sometimes the integration (looking at you, cheap zigbee devices and cloud dependent devices) doesn't manage to get the device into the desired state. This results in things like activating a scene and having only half the lights change. I think with this proposal it would be possible to automatically check if the actual device state reached the target state within some timeout and if not then retry. With that behavior unreliable devices would end up in the correct state, just more slowly, which IMO is a much better outcome for users than simply ignoring whether or not a device actually changed state as requested.

**Comment by jankornblum** [comment]
Any news regarding this?
